UserAgent/Home View/BookmarksView.swift

```diff
@@ -0,0 +1,240 @@
+            print("Source not set, aborting.")
+            return
+        }
+
+        // Don't delete folders, we don't create them anyway
+        guard !(bookmark is BookmarkFolder) else { return }
+
+        // Block to do this -- this is UI code.
+        guard let factory = source.modelFactory.value.successValue else {
+            print("Couldn't get model factory. This is unexpected.")
+            self.onModelFailure(DatabaseError(description: "Unable to get factory."))
+            return
+        }
+
+        let specificFactory = factory.factoryForIndex(indexPath.row, inFolder: source.current)
+        if let err = specificFactory.removeByGUID(bookmark.guid).value.failureValue {
+            print("Failed to remove \(bookmark.guid).")
+            self.onModelFailure(err)
+            return
+        }
+
+        self.source = source.removeGUIDFromCurrent(bookmark.guid)
+        self.updateEmptyPanelState()
+    }
+
+    override func pinToTopSites(_ site: Site) {
+        _ = self.profile.history.addPinnedTopSite(site).value
+    }
+
+    override func reloadData() {
+        self.source?.reloadData().upon(onModelFetched)
+    }
+
+    func loadData() {
+        // If we've not already set a source for this panel, fetch a new model from
+        // the root; otherwise, just use the existing source to select a folder.
+        guard let source = self.source else {
+            // Get all the bookmarks split by folders
+            if let bookmarkFolder = bookmarkFolder {
+                profile.bookmarks.modelFactory >>== { $0.modelForFolder(bookmarkFolder).upon(self.onModelFetched) }
+            } else {
+                profile.bookmarks.modelFactory >>== { $0.modelForRoot().upon(self.onModelFetched) }
+            }
+            return
+        }
+
+        if let bookmarkFolder = bookmarkFolder {
+            source.selectFolder(bookmarkFolder).upon(onModelFetched)
+        } else {
+            source.selectFolder(BookmarkRoots.MobileFolderGUID).upon(onModelFetched)
+        }
+    }
+
+    fileprivate func onModelFetched(_ result: Maybe<BookmarksModel>) {
+        guard let model = result.successValue else {
+            self.onModelFailure(result.failureValue as Any)
+            return
+        }
+        self.onNewModel(model)
+    }
+
+    fileprivate func onNewModel(_ model: BookmarksModel) {
+        if Thread.current.isMainThread {
+            self.source = model
+            self.tableView.reloadData()
+            return
+        }
+
+        DispatchQueue.main.async {
+            self.source = model
+            self.tableView.reloadData()
+            self.updateEmptyPanelState()
+        }
+    }
+
+    fileprivate func onModelFailure(_ e: Any) {
+        print("Error: failed to get data: \(e)")
+    }
+
+}
+
+// MARK: - Private API
+private extension BookmarksView {
+    private func updateEmptyPanelState() {
+        // TOOD
```

Comment:
As we decided to create new ticket for Todos. Please create new ticket for this todo.

Reply:
Whoops you caught me. It wasn't highlighted because of the typo. THanks

![](https://media.giphy.com/media/KJ1f5iTl4Oo7u/giphy.gif)